vembrane/modules/sort.py

```diff
@@ -0,0 +1,138 @@
+    )
+    parser.add_argument(
+        "--output",
+        "-o",
+        default="-",
+        help="Output file, if not specified, output is written to STDOUT.",
+    )
+    parser.add_argument(
+        "--output-fmt",
+        "-O",
+        default="vcf",
+        choices=["vcf", "bcf", "uncompressed-bcf"],
+        help="Output format.",
+    )
+    add_common_arguments(parser)
+
+
+def execute(args) -> None:
+    overwrite_number = {
+        "INFO": dict(args.overwrite_number_info),
+        "FORMAT": dict(args.overwrite_number_format),
+    }
+    fmt = {"vcf": "", "bcf": "b", "uncompressed-bcf": "u"}[args.output_fmt]
+
+    try:
+        with create_reader(
+            args.vcf,
+            backend=args.backend,
+            overwrite_number=overwrite_number,
+        ) as reader:
+            # TODO: unify the reader/writer setup across subcommands
+            reader.header.add_generic("vembraneVersion", __version__)
+            reader.header.add_generic(
+                "vembraneCmd",
+                "vembrane "
+                + " ".join(
+                    normalize(arg) if " " in arg else arg for arg in sys.argv[1:]
+                ),
+            )
+
+            annotation = Annotation(args.annotation_key, reader.header)
+
+            # load records into memory
+            # since sorting by something else than the position is only reasonable
+            # for already small, filtered VCF files, this is not an issue in practice.
+            # TODO: auto-detect larger files and use a mergesort library for those?
+            records = list(enumerate(reader))
+
+            sort_keys = SourceEnvironment(
+                args.expression, args.annotation_key, reader.header
+            )
+
+        def get_sort_key(item):
+            idx, record = item
+            sort_keys.update_from_record(idx, record)
+
+            def handle_na(value) -> Any:
+                # ensures that NA/None values are sorted to the end
+                if isinstance(value, Iterable) and not isinstance(value, str):
+                    # if the value is a tuple, we need to handle each element
+                    return tuple(handle_na(v) for v in value)
+                if (
+                    value is NA
+                    or value is None
+                    or (isinstance(value, float) and math.isnan(value))
+                ):
+                    return NAToEnd()
+                return value
+
+            if sort_keys.expression_annotations():
+                # if the expression refers to annotations, we obtain the
+                # minimal keys across all annotations
+                annotations = annotation.get_record_annotations(idx, record)
+
+                def eval_with_ann(annotation):
+                    sort_keys.update_annotation(annotation)
+                    return handle_na(eval(sort_keys.compiled, sort_keys))
+
+                return min(map(eval_with_ann, annotations))
```

Comment:
So annotations aren't sorted within the list of annotations, just the records based on the "smallest" ann? Is that mentioned / explained somewhere? How much work would it be to do that? Because at least `filter` also operates on that level (filtering out individual annotations within a single record)

Reply:
Yes, I think there is no use case for sorting annotations. The purpose is to get the records in the desired order. But, when combining this e.g. with table, it would be a bit weird if the annotations then are sorted within each record. I wonder what would be the best strategy here...

Reply:
At the moment, I even favor the splitting strategy over the min strategy.

Reply:
So, everything as discussed now. The code automatically uses external sorting in case the VCF exceeds a given size. Also, we have `asc()` and `desc()` decorators for keys that allow to set sort order per tuple item. The default is `asc()` so that it can be omitted for brevity.

Reply:
Further, we now have a flag --preserve-annotation-order that disables the default behavior (sorting annotations as well).

Reply:
In other words, ready for review @tedil!